logdetective/server/server.py

```diff
@@ -90,10 +94,45 @@ async def lifespan(fapp: FastAPI):
     # Ensure that the database is initialized.
     logdetective.server.database.base.init()
 
+    # Set up a queue for LLM requests.
+    # When talking to an LLM service, there's a limited number of requests
+    # we can make per-minute.
+    fapp.llm_queue = asyncio.Queue(maxsize=SERVER_CONFIG.inference.max_queue_size)
+    asyncio.create_task(process_queue(fapp.llm_queue))
+
     yield
+
+    # We're shutting down, so cancel any remaining items in the queue
+    fapp.llm_queue.shutdown(immediate=True)
     await fapp.http.close()
 
 
+def enqueue_func(queue, coro, *args, **kwargs):
+    loop = asyncio.get_running_loop()
+    future = loop.create_future()
+
+    queue.put_nowait((future, coro, args, kwargs))
+    return future
+
+
+async def process_queue(llm_queue):
+    """Run an item on the queue every interval seconds.
+
+    """
+    while True:
+        try:
+            future, coro, args, kwargs = await llm_queue.get()
+            result = await coro(*args, **kwargs)
```

Comment:
I wonder if we could process these in parallel and make that option configurable
i.e. process 3 coroutines in parallel

Reply:
agreed, that would be amazing

I hope we can agree on some configuration in the coming days and deploy it live

shall we start with `sleep 0.1s` and unlimited request count, and see how the servers take that?